rpm/build/SPECS/jenkins.spec

```diff
@@ -45,103 +42,42 @@ rm -rf "%{buildroot}"
-%__sed -i 's,~~WAR~~,%{_javadir}/%{name}.war,g' "%{buildroot}/etc/init.d/%{name}"
-%__install -d "%{buildroot}/usr/sbin"
-%__ln_s "../../etc/init.d/%{name}" "%{buildroot}/usr/sbin/rc%{name}"
-
-%__install -D -m0600 "%{SOURCE2}" "%{buildroot}/etc/sysconfig/%{name}"
-%__sed -i 's,~~HOME~~,%{workdir},g' "%{buildroot}/etc/sysconfig/%{name}"
-
-%__install -D -m0644 "%{SOURCE3}" "%{buildroot}/etc/logrotate.d/%{name}"
+%__install -D -m0644 "%{SOURCE3}" "%{buildroot}%{_sysconfdir}/logrotate.d/%{name}"
 
 %__install -D -m0644 "%{SOURCE4}" "%{buildroot}%{_unitdir}/%{name}.service"
+
 %__install -D -m0755 "%{SOURCE5}" "%{buildroot}%{_bindir}/%{name}"
-%__install -d "%{buildroot}%{_datadir}/%{name}"
-%__install -D -m0755 "%{SOURCE6}" "%{buildroot}%{_datadir}/%{name}/migrate"
 
 %pre
 /usr/bin/getent group %{name} &>/dev/null || /usr/sbin/groupadd -r %{name} &>/dev/null
 # SUSE version had -o here, but in Fedora -o isn't allowed without -u
 /usr/bin/getent passwd %{name} &>/dev/null || /usr/sbin/useradd -g %{name} -s /bin/false -r -c "@@SUMMARY@@" \
 	-d "%{workdir}" %{name} &>/dev/null
 
-  # Used to decide later if we should perform a chown in case JENKINS_INSTALL_SKIP_CHOWN is false
-  # Check if a previous installation exists, if so check the JENKINS_HOME value and existing owners of work, log and cache dir, need to to this check
-  # here because the %files directive overwrites folder owners, I have not found a simple way to make the
-  # files directive to use JENKINS_USER as owner.
-  if [ -f "/etc/sysconfig/%{name}" ]; then
-      logger -t %{name}.installer "Found previous config file /etc/sysconfig/%{name}"
-      . "/etc/sysconfig/%{name}"
-      stat --format=%U "/var/cache/%{name}" > "/tmp/%{name}.installer.cacheowner"
-      stat --format=%U "/var/log/%{name}"  >  "/tmp/%{name}.installer.logowner"
-      stat --format=%U ${JENKINS_HOME:-%{workdir}}  > "/tmp/%{name}.installer.workdirowner"
-  else
-      logger -t %{name}.installer "No previous config file /etc/sysconfig/%{name} found"
-  fi
-
 %post
-%{_datadir}/%{name}/migrate "/etc/sysconfig/%{name}" || true
 %systemd_post %{name}.service
 
-function chownIfNecessary {
-  logger -t %{name}.installer "Checking ${2} ownership"
-  if [ -f "${1}" ] ; then
-    owner=$(cat "$1")
-    rm -f "$1"
-    logger -t %{name}.installer "Found previous owner of ${2}: ${owner} "
-  fi
-  if  [ "${owner:-%{name}}" != "${JENKINS_USER:-%{name}}" ] ; then
-    logger -t %{name}.installer "Previous owner of ${2} is different than configured JENKINS_USER... Doing a recursive chown of ${2} "
-    chown -R ${JENKINS_USER:-%{name}} "$2"
-  elif [ "${JENKINS_USER:-%{name}}" != "%{name}" ] ; then
-    # User has changed ownership of files and JENKINS_USER, chown only the folder
-    logger -t %{name}.installer "Configured JENKINS_USER is different than default... Doing a non recursive chown of ${2} "
-    chown ${JENKINS_USER:-%{name}} "$2"
-  else
-    logger -t %{name}.installer "No chown needed for ${2} "
-  fi
-}
```

Comment:
I'm a little worried about the outright deletion of this logic. Suppose a user of the existing package is running with a custom user and group, and then they upgrade to a new version with your changes. Have you explored this scenario logically as well as with empirical testing? I am afraid that this test would fail. There is no automated test for this. In fact, the same bug was present in the Debian version: https://github.com/jenkinsci/packaging/pull/344

Reply:
It's a bad style to re-chown/chgrp packaged files:
```bash
chown nobody: /var/lib/jenkins
rpm -Vv --noconfig jenkins | grep -vE '^\.{9}'
# .....UG..    /var/lib/jenkins
```
When running jenkins under a different user/group, a different working directory (that's not included in the rpm) should be used. And this different working directory would be reflected in `/etc/systemd/system/jenkins.service.d/override.conf` (at least it should).
But you're right, this would be a breaking change. I'll add a workaround to the specfile in the next days.

Reply:
How about this?
```diff
--- a/rpm/build/SPECS/jenkins.spec
+++ b/rpm/build/SPECS/jenkins.spec
@@ -58,9 +58,17 @@ rm -rf "%{buildroot}"
 	-d "%{workdir}" %{name} &>/dev/null
 
 %post
+if [ $1 -eq 1 ]; then
+    %__install -d -m 0755 -o %{name} -g %{name} %{workdir}
+    %__install -d -m 0750 -o %{name} -g %{name} %{_localstatedir}/log/%{name}
+    %__install -d -m 0750 -o %{name} -g %{name} %{_localstatedir}/cache/%{name}
+fi
 %systemd_post %{name}.service
 
 %preun
+if [ $1 -eq 0 ]; then
+    %__rm -rf %{_localstatedir}/cache/%{name}/war
+fi
 %systemd_preun %{name}.service
 
 %postun
@@ -68,9 +76,9 @@ rm -rf "%{buildroot}"
 
 %files
 %{_javadir}/%{name}.war
-%attr(0755,%{name},%{name}) %dir %{workdir}
-%attr(0750,%{name},%{name}) %{_localstatedir}/log/%{name}
-%attr(0750,%{name},%{name}) %{_localstatedir}/cache/%{name}
+%ghost %{workdir}
+%ghost %{_localstatedir}/log/%{name}
+%ghost %{_localstatedir}/cache/%{name}
 %config %{_sysconfdir}/logrotate.d/%{name}
 %{_unitdir}/%{name}.service
 %{_bindir}/%{name}

```
When installing the package for the first time, the directories `/var/lib/jenkins`, `/var/log/jenkins` and `/var/cache/jenkins` are created with the corresponding attributes. Since the directories are marked as `%ghost` no changes are made to the directories on package upgrades. Upon deinstallation, the `/var/cache/jenkins/war` directory gets removed to leave no temporary data behind.

Reply:
That sounds about right to me. I wonder if we really need special handling for `/var/log/jenkins`. While the default systemd configuration does put `JENKINS_HOME` in `/var/lib/jenkins` and does put the exploded WAR into `/var/cache/jenkins/war`, it does _not_ (anymore) put the log in `/var/log/jenkins`. Rather `systemd-journald(8)` is used and this line is commented out by default:

```
#Environment="JENKINS_LOG=%L/@@ARTIFACTNAME@@/@@ARTIFACTNAME@@.log"
```

One could argue that if a user is moving from `systemd-journald(8)` to a log file by uncommenting this line, then it is their responsibility to create the log directory and ensure it has appropriate permissions. And in the default configuration, where `systemd-journald(8)` is being used, it might be confusing and unnecessary for that empty directory to be created when installing the package.

I don't feel strongly about this either way, though. It's possible I'm missing something.

Reply:
Agreed that we can drop the logrotate config. Happy if you want to do that in this PR, or it can be deferred to a separate issue/PR if desired.

Reply:
:+1: I removed the logrotate config as well.
The specfile shrank quite a lot with all these changes.